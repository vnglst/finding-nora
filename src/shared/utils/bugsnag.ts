import bugsnag from "@bugsnag/js";
import bugsnagReact from "@bugsnag/plugin-react";
import * as React from "react";

export const bugsnagClient = bugsnag("eb9c66e47f7f95c5801a21ffe1308619");
bugsnagClient.use(bugsnagReact, React);
export const BugsnagErrorBoundary = bugsnagClient.getPlugin("react");
