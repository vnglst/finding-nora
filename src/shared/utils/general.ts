export const sample = (arr: any[]) =>
  arr[Math.floor(Math.random() * arr.length)]

// What the compose below does is basically this:
// const compose = (a: any, b: any) => (x: any) => a(b(x))
// but then for any number of functions
export const compose = (...functions: any[]) => (data: any) =>
  functions.reduceRight((value, func) => func(value), data)

export const cloneDeep = compose(
  JSON.parse,
  JSON.stringify
)
