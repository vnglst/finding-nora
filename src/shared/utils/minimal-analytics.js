export function initialize(context, trackingId, options) {
  const history = context.history
  const doc = document
  const nav = navigator || {}
  const storage = localStorage
  const encode = encodeURIComponent
  const pushState = history.pushState
  const typeException = 'exception'
  const generateId = () => Math.random().toString(36)
  const getId = () => {
    if (!storage.cid) {
      storage.cid = generateId()
    }
    return storage.cid
  }
  const serialize = obj => {
    var str = []
    for (var p in obj) {
      if (obj.hasOwnProperty(p)) {
        if (obj[p] !== undefined) {
          str.push(encode(p) + '=' + encode(obj[p]))
        }
      }
    }
    return str.join('&')
  }
  const track = (
    type,
    eventCategory,
    eventAction,
    eventLabel,
    eventValue,
    exceptionDescription,
    exceptionFatal
  ) => {
    const url = 'https://www.google-analytics.com/collect'
    const data = serialize({
      v: '1',
      ds: 'web',
      aip: options.anonymizeIp ? 1 : undefined,
      tid: trackingId,
      cid: getId(),
      t: type || 'pageview',
      sd:
        options.colorDepth && context.screen.colorDepth
          ? `${context.screen.colorDepth}-bits`
          : undefined,
      dr: doc.referrer || undefined,
      dt: doc.title,
      dl: doc.location.origin + doc.location.pathname + doc.location.search,
      ul: options.language ? (nav.language || '').toLowerCase() : undefined,
      de: options.characterSet ? doc.characterSet : undefined,
      sr: options.screenSize
        ? `${(context.screen || {}).width}x${(context.screen || {}).height}`
        : undefined,
      vp:
        options.screenSize && context.visualViewport
          ? `${(context.visualViewport || {}).width}x${
              (context.visualViewport || {}).height
            }`
          : undefined,
      ec: eventCategory || undefined,
      ea: eventAction || undefined,
      el: eventLabel || undefined,
      ev: eventValue || undefined,
      exd: exceptionDescription || undefined,
      exf:
        typeof exceptionFatal !== 'undefined' && !!exceptionFatal === false
          ? 0
          : undefined
    })

    if (nav.sendBeacon) {
      nav.sendBeacon(url, data)
    } else {
      var xhr = new XMLHttpRequest()
      xhr.open('POST', url, true)
      xhr.send(data)
    }
  }
  const trackEvent = (category, action, label, value) =>
    track('event', category, action, label, value)
  const trackException = (description, fatal) =>
    track(typeException, null, null, null, null, description, fatal)
  history.pushState = function(state) {
    if (typeof history.onpushstate == 'function') {
      history.onpushstate({ state: state })
    }
    setTimeout(track, options.delay || 10)
    return pushState.apply(history, arguments)
  }
  track()
  context.ma = {
    trackEvent,
    trackException
  }
}
