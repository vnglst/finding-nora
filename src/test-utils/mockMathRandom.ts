export const mockMathRandom = () => {
  const mockGlobalMathLibrary = Object.create(global.Math)
  mockGlobalMathLibrary.random = generatePseudoRandomNumberWithSeed(42)
  global.Math = mockGlobalMathLibrary
}

const generatePseudoRandomNumberWithSeed = (seed: number) => {
  const generatePseudoRandomNumber = () => {
    const pseudoRandomNumber = Math.sin(seed++) * 10000
    return pseudoRandomNumber - Math.floor(pseudoRandomNumber)
  }
  return generatePseudoRandomNumber
}
