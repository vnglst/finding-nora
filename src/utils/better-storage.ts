export const localStore = storageFactory(localStorage);
export const sessionStore = storageFactory(sessionStorage);

// source: https://gist.github.com/MichalZalecki/70a831304b0e83afbc0fbad3e8577068
export function storageFactory(storage: Storage): Storage {
  let inMemoryStorage: { [key: string]: string } = {};
  const length = 0;

  function isSupported() {
    try {
      const testKey = "__some_random_key_to_test_if_storage_is_supported__";
      storage.setItem(testKey, testKey);
      storage.removeItem(testKey);
      return true;
    } catch (e) {
      return false;
    }
  }

  function clear(): void {
    if (isSupported()) {
      storage.clear();
    } else {
      inMemoryStorage = {};
    }
  }

  function getItem(name: string): string | null {
    if (isSupported()) {
      return storage.getItem(name);
    }

    if (Object.prototype.hasOwnProperty.call(inMemoryStorage, name)) {
      return inMemoryStorage[name];
    }
    return null;
  }

  function key(index: number): string | null {
    if (isSupported()) {
      return storage.key(index);
    } else {
      return Object.keys(inMemoryStorage)[index] || null;
    }
  }

  function removeItem(name: string): void {
    if (isSupported()) {
      storage.removeItem(name);
    } else {
      delete inMemoryStorage[name];
    }
  }

  function setItem(name: string, value: string): void {
    if (isSupported()) {
      storage.setItem(name, value);
    } else {
      inMemoryStorage[name] = String(value);
    }
  }

  return {
    getItem,
    setItem,
    removeItem,
    clear,
    key,
    length
  };
}
