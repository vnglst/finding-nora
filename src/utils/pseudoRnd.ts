export function createPseudoRnd(seed = 42) {
  const pseudoRnd = () => {
    const pseudoRandomNumber = Math.sin(seed++) * 10000;
    return pseudoRandomNumber - Math.floor(pseudoRandomNumber);
  };
  return pseudoRnd;
}
