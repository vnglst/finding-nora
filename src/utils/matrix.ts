// inspired by this article:
// https://medium.com/front-end-hacking/matrix-rotation-%EF%B8%8F-6550397f16ab

import { cloneDeep, compose } from './general'

const flipMatrix = (matrix: any[][]) =>
  matrix[0].map((column: any, index) => matrix.map((row: any) => row[index]))

export const mirrorMatrixHorizontally = (matrix: any[][]) =>
  cloneDeep(matrix).reverse()

export const mirrorMatrixVertically = (matrix: any[][]) =>
  cloneDeep(matrix).map((row: any) => row.reverse())

export const rotateMatrixCounterClockwise = compose(
  mirrorMatrixHorizontally,
  flipMatrix,
)

export const rotateMatrixClockwise = compose(
  flipMatrix,
  mirrorMatrixHorizontally,
)
