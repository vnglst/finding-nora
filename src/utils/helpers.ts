import { createPseudoRnd } from "./pseudoRnd";

// Use a pseudo random function with seed number in cypress test
// for repeatable test scenarios in unit + e2e tests
const usePseudoRnd = process.env.REACT_APP_PSEUDO_RND === "true";
const rnd = usePseudoRnd ? createPseudoRnd() : () => Math.random();

export function getRnd(max: number) {
  return Math.floor(rnd() * max);
}

export function shuffle<T>(arr: T[]): T[] {
  return arr.sort(() => rnd() - 0.5);
}

export function sample<T>(arr: T[]) {
  return arr[Math.floor(rnd() * arr.length)];
}

export function deepEqual<T>(obj1: T, obj2: T) {
  return JSON.stringify(obj1) === JSON.stringify(obj2);
}

export function cloneDeep<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj));
}
