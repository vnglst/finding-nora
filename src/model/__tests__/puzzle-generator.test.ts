import { mockMathRandom } from 'utils/mockMathRandom'
import PuzzleGenerator from '../puzzle-generator'

const noise = 'ABCDEFGHIJKLMNOPQRSTVWUXYZ'.split('')
const solution = 'NORA'.toUpperCase().split('')

beforeEach(() => {
  mockMathRandom()
})

describe('Puzzle', () => {
  it('Should generate a 6 x 6 grid', () => {
    const puzzle = new PuzzleGenerator(6, solution, noise)
    expect(puzzle.grid.length).toEqual(6)
    expect(puzzle.grid[0].length).toEqual(6)
  })
  it('Should contain all solution letters', () => {
    const puzzle = new PuzzleGenerator(4, solution, noise)
    solution.forEach(letter => {
      expect(puzzle.find(letter)).toBeTruthy()
    })
  })
  it('Should not contain solution letters twice', () => {
    const puzzle = new PuzzleGenerator(8, solution, noise)
    solution.forEach(letter => {
      expect(puzzle.find(letter).length).toBeLessThan(2)
    })
  })
  it('Should match snapshot for large puzzle', () => {
    const puzzle = new PuzzleGenerator(8, solution, noise)
    expect(puzzle.grid).toMatchSnapshot()
  })
})

describe('removeSolutionLettersFrom()', () => {
  const puzzle = new PuzzleGenerator(6, solution, noise)
  puzzle.removeSolutionLettersFrom('NORABCD'.split(''))
  expect(puzzle.noise).toEqual(['B', 'C', 'D'])
})

describe('calculateStepsToEdge()', () => {
  it('Should calculate steps to edge start at origin', () => {
    const puzzle = new PuzzleGenerator(6, solution, noise)
    const steps = puzzle.calculateStepsToEdge(0, 0)
    expect(steps).toEqual(11)
  })
  it('Should calculate steps to edge start at center', () => {
    const puzzle = new PuzzleGenerator(6, solution, noise)
    const steps = puzzle.calculateStepsToEdge(2, 2)
    expect(steps).toEqual(7)
  })
})

describe('generateLegalRandomStartPoint()', () => {
  it('Should generate a legal random startpoint', () => {
    const puzzle = new PuzzleGenerator(6, solution, noise)
    const startPoint = puzzle.generateLegalRandomStartPoint()
    expect(typeof startPoint.startColumn).toEqual('number')
    expect(typeof startPoint.startRow).toEqual('number')
  })
  it('Should throw error if puzzle cannot be added', () => {
    expect(() => {
      const puzzle = new PuzzleGenerator(2, solution, noise)
      puzzle.generateLegalRandomStartPoint()
    }).toThrowError('Puzzle too long')
  })
})

describe('updateIndices()', () => {
  it('Should update column and rows after changes', () => {
    const puzzle = new PuzzleGenerator(6, solution, noise)
    puzzle.grid[1][1].column = 2
    puzzle.grid[1][1].row = 2
    puzzle.updateIndices()
    expect(puzzle.grid[1][1].column).toEqual(1)
    expect(puzzle.grid[1][1].row).toEqual(1)
  })
})

describe('applyTransformations()', () => {
  it('Should do nothing on first tranformation', () => {
    const puzzle = new PuzzleGenerator(4, solution, noise)
    puzzle.applyRandomTransformation()
    const str = puzzle.toString()
    expect(str.startsWith('BLGA')).toBeTruthy()
  })

  it('Should rotate counterclockwise on second tranformation', () => {
    const puzzle = new PuzzleGenerator(4, solution, noise)
    puzzle.applyRandomTransformation()
    puzzle.applyRandomTransformation()
    const str = puzzle.toString()
    expect(str.startsWith('ARKD')).toBeTruthy()
  })

  it('Should mirror horizontally on third tranformation', () => {
    const puzzle = new PuzzleGenerator(4, solution, noise)
    puzzle.applyRandomTransformation()
    puzzle.applyRandomTransformation()
    puzzle.applyRandomTransformation()
    const str = puzzle.toString()
    expect(str.endsWith('ARKD\n')).toBeTruthy()
  })
})
