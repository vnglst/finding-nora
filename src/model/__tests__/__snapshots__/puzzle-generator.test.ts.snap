// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Puzzle Should match snapshot for large puzzle 1`] = `
Array [
  Array [
    Object {
      "column": 0,
      "letter": "H",
      "row": 0,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 1,
      "letter": "F",
      "row": 0,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 2,
      "letter": "N",
      "row": 0,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 3,
      "letter": "O",
      "row": 0,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 4,
      "letter": "R",
      "row": 0,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 5,
      "letter": "A",
      "row": 0,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 6,
      "letter": "S",
      "row": 0,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 7,
      "letter": "D",
      "row": 0,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
  ],
  Array [
    Object {
      "column": 0,
      "letter": "P",
      "row": 1,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 1,
      "letter": "D",
      "row": 1,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 2,
      "letter": "M",
      "row": 1,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 3,
      "letter": "U",
      "row": 1,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 4,
      "letter": "P",
      "row": 1,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 5,
      "letter": "X",
      "row": 1,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 6,
      "letter": "X",
      "row": 1,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 7,
      "letter": "C",
      "row": 1,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
  ],
  Array [
    Object {
      "column": 0,
      "letter": "E",
      "row": 2,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 1,
      "letter": "F",
      "row": 2,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 2,
      "letter": "F",
      "row": 2,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 3,
      "letter": "V",
      "row": 2,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 4,
      "letter": "P",
      "row": 2,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 5,
      "letter": "E",
      "row": 2,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 6,
      "letter": "X",
      "row": 2,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 7,
      "letter": "T",
      "row": 2,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
  ],
  Array [
    Object {
      "column": 0,
      "letter": "K",
      "row": 3,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 1,
      "letter": "D",
      "row": 3,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 2,
      "letter": "S",
      "row": 3,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 3,
      "letter": "F",
      "row": 3,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 4,
      "letter": "C",
      "row": 3,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 5,
      "letter": "J",
      "row": 3,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 6,
      "letter": "Q",
      "row": 3,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 7,
      "letter": "J",
      "row": 3,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
  ],
  Array [
    Object {
      "column": 0,
      "letter": "Q",
      "row": 4,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 1,
      "letter": "B",
      "row": 4,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 2,
      "letter": "S",
      "row": 4,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 3,
      "letter": "P",
      "row": 4,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 4,
      "letter": "W",
      "row": 4,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 5,
      "letter": "G",
      "row": 4,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 6,
      "letter": "K",
      "row": 4,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 7,
      "letter": "M",
      "row": 4,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
  ],
  Array [
    Object {
      "column": 0,
      "letter": "C",
      "row": 5,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 1,
      "letter": "B",
      "row": 5,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 2,
      "letter": "Y",
      "row": 5,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 3,
      "letter": "L",
      "row": 5,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 4,
      "letter": "M",
      "row": 5,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 5,
      "letter": "E",
      "row": 5,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 6,
      "letter": "X",
      "row": 5,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 7,
      "letter": "C",
      "row": 5,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
  ],
  Array [
    Object {
      "column": 0,
      "letter": "Z",
      "row": 6,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 1,
      "letter": "F",
      "row": 6,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 2,
      "letter": "P",
      "row": 6,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 3,
      "letter": "Z",
      "row": 6,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 4,
      "letter": "M",
      "row": 6,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 5,
      "letter": "I",
      "row": 6,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 6,
      "letter": "G",
      "row": 6,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 7,
      "letter": "Q",
      "row": 6,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
  ],
  Array [
    Object {
      "column": 0,
      "letter": "E",
      "row": 7,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 1,
      "letter": "C",
      "row": 7,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 2,
      "letter": "B",
      "row": 7,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 3,
      "letter": "H",
      "row": 7,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 4,
      "letter": "J",
      "row": 7,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 5,
      "letter": "Q",
      "row": 7,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 6,
      "letter": "K",
      "row": 7,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
    Object {
      "column": 7,
      "letter": "X",
      "row": 7,
      "updatedAt": "2017-07-14T02:40:00.000Z",
    },
  ],
]
`;
